src/librustc_typeck/check/mod.rs

```diff
@@ -1859,23 +1873,31 @@ fn check_transparent<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, sp: Span, def_id: De
     if non_zst_count != 1 {
         let field_spans: Vec<_> = non_zst_fields.map(|(span, _zst, _align1)| span).collect();
 
-        let mut err = struct_span_err!(tcx.sess, sp, E0690,
-                         "{}transparent {} needs exactly one non-zero-sized field, but has {}",
-                         if adt.is_enum() { "the variant of a " } else { "" },
-                         adt.descr(),
-                         non_zst_count);
-        if !field_spans.is_empty() {
-            err.span_note(field_spans,
-                          &format!("the following non-zero-sized fields exist on `{}`:",
-                                   tcx.def_path_str(def_id)));
+        let msg = format!("needs exactly one non-zero-sized field, but has {}", non_zst_count);
+        let mut err = struct_span_err!(
+            tcx.sess,
+            sp,
+            E0690,
+            "{}transparent {} {}",
+            if adt.is_enum() { "the variant of a " } else { "" },
+            adt.descr(),
+            msg,
+        );
+        err.span_label(sp, &msg);
+        for sp in &field_spans {
+            err.span_label(*sp, "this field is non-zero-sized");
         }
         err.emit();
     }
     for (span, zst, align1) in field_infos {
         if zst && !align1 {
-            span_err!(tcx.sess, span, E0691,
-                      "zero-sized field in transparent {} has alignment larger than 1",
-                      adt.descr());
+            struct_span_err!(
```

Comment:
and here.

Reply:
This feels small enough for it to be reasonable staying there ^_^

Reply:
fair :)